doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,401 @@
+  data is either written twice (WAL) or written in a log-structured manner.
+
+  (1) If Request Size ≤ Threshold (similar to minimum allocation size in BlueStore)
+    Write data and metadata to [WAL] —flush—> Write them to [Data section (in-place)] and 
+    [Metadata section], respectively.
+
+    Since the CPU becomes the bottleneck for small I/O workloads, in-place update scheme is used.
+    Double write for small data may be better than host-GC overhead in terms of CPU consumption 
+    in the long run
+
+  (2) Else if Request Size > Threshold
+    Append data to [Data section (log-structure)] —> Write the corresponding metadata to [WAL] 
+    —flush—> Write the metadata to [Metadata section]
+
+    For large I/O workloads, the host can afford process host-GC
+    Also, the host can garbage collect invalid objects more effectively when their size is large
+
+    Note that Threshold can be configured to a very large number so that only the scenario (1) occurs.
+    With this design, we can control the overall I/O procedure with the optimizations for crimson
+    as described above.
+
+  - Detailed flow
+    - 1. Append a log entry that contains pg_log, snapset, object_infot_t, block allocation
+      using NVMe atomic write command on the WAL
+      - NVMe provides atomicity guarantees for a write command (Atomic Write Unit Power Fail)
+        For example, 512 Kbytes of data can be atomically written at once without fsync()
+      - Small size (object_info_t, snapset, collection, etc.) can be embed
+      - stage 1
+        WAL (written) --> | TxBegin A | Log Entry | TxEnd A | (if small)
+        Data partition (written)--> | Data blocks | (if large)
+    - 2. Then, append the data to WAL (if large)
+      - stage 2 (Updating object meta and allocation bitmap can be skipped via pre-allocation)
+        WAL --> | TxBegin A | Log Entry | TxEnd A | (if large)
+
+- Read
+  - Use the cached object metadata to find out the data location
+  - If not cached, need to search WAL after checkpoint and Object meta partition to find the 
+    latest meta data
+
+- Flush
+  - Flush WAL entries whenever needed
+
+
+Crash consistency
+------------------
+Large case
+- 1. Crash occurs right after writing Data blocks
+  - Data partition --> | Data blocks |
+  - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+- 2. Crash occurs right after WAL using atomic write command
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - Data partition --> | Data blocks |
+  - Write procedure is completed, so there is no data loss or inconsistent state
+
+Small case
+- 1. Crash occurs right after writing WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need two writes on both WAL and Data partition without updating Object meta and Data bitmap
+- Worst case (flush happens)
```

Comment:
"flush happens" I don't understand.  Do you mean that you don't have enough preallocated usable space to do the data write?